tower-util/src/combinators/map_err.rs

```diff
@@ -0,0 +1,72 @@
+use futures_util::{future::MapErr as MapErrFut, TryFutureExt};
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+#[derive(Debug)]
+pub struct MapErr<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapErr<S, F> {
+    pub fn new(inner: S, f: F) -> Self {
+        MapErr { f, inner }
+    }
+}
+
+impl<S, F, Request, Error> Service<Request> for MapErr<S, F>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error,
+    F: Clone,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = MapErrFut<S::Future, F>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(self.f.clone())
+    }
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        self.inner.call(request).map_err(self.f.clone())
+    }
+}
+
+#[derive(Debug)]
+pub struct MapErrLayer<F, Request, Error> {
+    f: F,
+    request: PhantomData<Request>,
+    error: PhantomData<Error>,
```

Comment:
nit: a few notes on the use of `PhantomData`:

1. let's use `PhantomData<fn(T)>` rather than `PhantomData<T>`, so that the type is not considered to contain a field of type `T`. that way, if `T` is not `Send` or `Sync`, the containing type can still implement those traits (since it doesn't actually contain a `T`).
2. since these fields are unused, the compiler will warn if we don't prefix them with `_` --- let's do that.
3. i'd prefer to collapse these into one `PhantomData` field that contains both the `Request` and `Error` types --- multiple `PhantomData`s just adds noise IMO.

so, in sum, i'd prefer something like this:
```suggestion
    _p: PhantomData<fn(Request, Error)>,
```